qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -137,99 +160,239 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
+            TranspilerError: if a 'method' was specified for the class and is not
+                found in the installed plugins list. The list of installed
+                plugins can be queried with
+                :func:`~qiskit.transpiler.passes.synthesis.plugins.unitary_synthesis_plugin_names`
         """
+        if not self.method:
+            method = "default"
+        else:
+            method = self.method
+        if method not in self.plugins.ext_plugins:
+            raise TranspilerError("Specified method: %s not found in plugin list" % method)
+        default_method = self.plugins.ext_plugins["default"].obj
+        plugin_method = self.plugins.ext_plugins[method].obj
+        if plugin_method.supports_coupling_map:
+            dag_bit_indices = {bit: idx for idx, bit in enumerate(dag.qubits)}
+        kwargs = {}
+        if plugin_method.supports_basis_gates:
+            kwargs["basis_gates"] = self._basis_gates
+        if plugin_method.supports_natural_direction:
+            kwargs["natural_direction"] = self._natural_direction
+        if plugin_method.supports_pulse_optimize:
+            kwargs["pulse_optimize"] = self._pulse_optimize
+        if plugin_method.supports_gate_lengths:
+            kwargs["gate_lengths"] = _build_gate_lengths(self._backend_props)
+        if plugin_method.supports_gate_errors:
+            kwargs["gate_errors"] = _build_gate_errors(self._backend_props)
+        supported_basis = plugin_method.supported_basis
+        if supported_basis is not None:
+            kwargs["matched_basis"] = _choose_basis(self._basis_gates, supported_basis)
+
+        # Handle approximation degree as a special case for backwards compatibility, it's
+        # not part of the plugin interface and only something needed for the default
+        # pass.
+        default_method._approximation_degree = self._approximation_degree
+        if method == "default":
+            plugin_method._approximation_degree = self._approximation_degree
+
+        for node in dag.named_nodes(*self._synth_gates):
+            if plugin_method.supports_coupling_map:
+                kwargs["coupling_map"] = self._coupling_map
+                kwargs["qubits"] = [dag_bit_indices[x] for x in node.qargs]
+            synth_dag = None
+            unitary = node.op.to_matrix()
+            n_qubits = len(node.qargs)
+            if (plugin_method.max_qubits is not None and n_qubits > plugin_method.max_qubits) or (
+                plugin_method.min_qubits is not None and n_qubits < plugin_method.min_qubits
+            ):
+                synth_dag = default_method.run(unitary, **kwargs)
+            else:
+                synth_dag = plugin_method.run(unitary, **kwargs)
+            if synth_dag:
+                if isinstance(synth_dag, tuple):
+                    dag.substitute_node_with_dag(node, synth_dag[0], wires=synth_dag[1])
+                else:
+                    dag.substitute_node_with_dag(node, synth_dag)
```

Comment:
If the return value is false-y (can a `DAGCircuit` be false-y?) then this wouldn't substitute the node at all, even if the value isn't `None`.  Should the test be `if synth_dag is not None`, so we can loudly fail if passed `False`, or not accidentally do something weird if the substitution value evaluates False?  (Not that I think the latter is a likely scenario.)

Reply:
I don't think a `DAGCircuit` could be false-y it doesn't implement `__bool__` but making this explicit is a good idea.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/6124/commits/5d8539ac37300473270b470574aa8879c975e2f9